### PR TITLE
Autotune Unit tests additional tests for all features

### PR DESCRIPTION
**Pull request by Philoul**
AutotuneCore:
=> Test4 added (dataset created with Autotune query from NS, test Autotune Core with PumpProfile, InputProfile and limiting basal and ISF with AutosensMax value)

AutotunePrep:
=> test2 and test3 added (dataset created from test1, with all carbs removed from treatments to test FullUAM and Categorize UAM as basal parameter On and Off)

Note, I had to fix Unit Test because prepped data for ISF and basal was not sorted exactly the same way (differences between Linux and kotlin...), Results identical with a sort by date included in unit test...

I think now all features are tested and Ok for Autotune Algo...

### CODE REVIEW

**Comment by Philoul** [comment]
I'll try to find what's wrong with tests (it passed on my computer...)

**Comment by MilosKozak** [comment]
FAILING tests

**Comment by Philoul** [comment]
Yes I'm trying to understand, but it's very strange... 
I had difficulties with sort method different on my computer and Linux.
I added a sort by date for ISF and basal categorrized data in AutotunePrepTest to fix it on my computer but it does not work on automated test

**Comment by Philoul** [comment]
I'll continue to try what could be wrong, but it's very complicated to analyse because everything works well on my computer...
How can I add log in dotests (I only see Assert information, I tryed aapslogger.debug but doesn't see log information in dotests...)

**Comment by Philoul** [comment]
@MilosKozak I need some hints to be able to debug. I'm blind without log message in dotests (only Assert result is visible).
I first thought it was a problem of sort, but the size of AAPS result list is not the same on my computer and on dotests...

**Comment by MilosKozak** [comment]
use `val aapsLogger = AAPSLoggerTest()` and log
time and timezone is common issue

File: app/src/test/java/info/nightscout/androidaps/plugins/general/autotune/AutotunePrepTest.kt [hunk 1]
```diff
@@ -87,10 +89,93 @@ class AutotunePrepTest : TestBaseWithProfile() {
...
+        `when`(sp.getDouble(R.string.key_openapsama_min_5m_carbimpact, 3.0)).thenReturn(min5mCarbImpact)
+        `when`(sp.getBoolean(R.string.key_autotune_categorize_uam_as_basal, false)).thenReturn(false)           // CategorizeUAM as Basal = False
+        val aapsPreppedGlucose = autotunePrep.categorizeBGDatums(inputProfile, inputProfile.localInsulin, false)
+        try {
+            aapsPreppedGlucose?.let {       // compare all categorization calculated by aaps plugin (aapsPreppedGlucose) with categorization calculated by OpenAPS (oapsPreppedGlucose)
+                for (i in aapsPreppedGlucose.crData.indices)
+                    Assert.assertTrue(oapsPreppedGlucose.crData[i].equals(aapsPreppedGlucose.crData[i]))
+                for (i in aapsPreppedGlucose.csfGlucoseData.indices)
+                    Assert.assertTrue(oapsPreppedGlucose.csfGlucoseData[i].equals(aapsPreppedGlucose.csfGlucoseData[i]))
+                aapsLogger.debug("Oaps ISF Size ${oapsPreppedGlucose.isfGlucoseData.size} / Aaps ISF Size ${aapsPreppedGlucose.isfGlucoseData.size}")
```
Comment (Philoul) [inline app/src/test/java/info/nightscout/androidaps/plugins/general/autotune/AutotunePrepTest.kt:116]: I already tried to use aapsLogger here, but it didn't help me (I can't see log information in CircleCI).
All these tests are Ok on my computer, and I can't reproduce what happens during automated dotests...

```
<============-> 99% EXECUTING [49s]> :app:testFullDebugUnitTest > Executing test info.nightscout.androidaps.interf> :app:testFullDebugUnitTest > Executing test info.nightscout...general.autotun> :app:testFullDebugUnitTest > 12 tests completed> :app:testFullDebugUnitTest > Executing test info.nightscout.androidaps.plugin<============-> 99% EXECUTING [50s]
info.nightscout.androidaps.plugins.general.autotune.AutotunePrepTest > autotunePrepTest2 FAILED
    java.lang.AssertionError: expected:<114> but was:<115>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:120)
        at org.junit.Assert.assertEquals(Assert.java:146)
        at info.nightscout.androidaps.plugins.general.autotune.AutotunePrepTest.autotunePrepTest2(AutotunePrepTest.kt:117)
```

**Comment by Philoul** [comment]
Not sure, latest commits was not for the unit tests but more to understand the differences between run on my computer and automated run.
I miss log information on run executed in CircleCI machine, so I modified unit test  with assert.equals and strings to get a little bit more information to help understanding why it's different.
It's frustrating to have tests that are 100% ok locally where you can analyze everything, and bad on the machine where no information is available.

**Comment by MilosKozak** [comment]
see https://github.com/nightscout/AndroidAPS/pull/1860

**Comment by Philoul** [comment]
Thank's, I'll take a look this evening. 👍

**Pull request closed without merging**